diff --git a/include/openssl/opensslv.h b/include/openssl/opensslv.h
index 7634eb14cd..cacf3e2ef1 100644
--- a/include/openssl/opensslv.h
+++ b/include/openssl/opensslv.h
@@ -130,11 +130,11 @@ const char *OPENSSL_version_build_metadata(void);
 #  else
 #   define _OPENSSL_VERSION_PRE_RELEASE 0xf
 #  endif
-#  define OPENSSL_VERSION_NUMBER        \
-    (long)( (OPENSSL_VERSION_MAJOR<<28)  \
+#  define OPENSSL_VERSION_NUMBER        \
+          ( (OPENSSL_VERSION_MAJOR<<28)  \
             |(OPENSSL_VERSION_MINOR<<20) \
             |(OPENSSL_VERSION_PATCH<<4)  \
-            |_OPENSSL_VERSION_PRE_RELEASE )
+            |_OPENSSL_VERSION_PRE_RELEASE )
 # endif
 
 # ifdef  __cplusplus
